File: sdk/node/src/profilesource.ts
// Fetching personas from the profile service, and caching them locally.
//
// Selection runs CLIENT-SIDE (see profilelib.ts) because it needs facts only this machine has:
// the real GPU, the real display, the engine's actual Chromium major. The server cannot see any
// of those, so it returns candidates for a described host and the SDK picks among them.
//
// The service caps DISTINCT profiles per license per day, and re-fetching one already held is
// free. That is why this module caches aggressively on disk: a sticky identity must resolve to
// the same profile every launch without spending budget or requiring the network.

import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { PROFILE_DIR } from "./profile.js";
import {
  selectProfile,
  gpuVendorClass,
  type HostFacts,
  type ProfileIndexEntry,
  type SelectOptions,
  type Selection,
} from "./profilelib.js";

const DEFAULT_API_BASE = "https://www.clearcotelabs.com";

function apiBase(base?: string): string {
  return (base || process.env.CLEARCOTE_PROFILE_API || process.env.CLEARCOTE_LICENSE_API || DEFAULT_API_BASE)
    .replace(/\/$/, "");
}

function cacheDir(): string {
  const d = join(PROFILE_DIR, "library");
  if (!existsSync(d)) mkdirSync(d, { recursive: true });
  return d;
}

export interface ProfileSourceOptions extends SelectOptions {
  licenseKey?: string;
  apiBase?: string;
  /** Skip the network entirely and use only what is already cached. */
  offline?: boolean;
}

/** Map a Node platform string onto the os_family values the corpus stores. */
export function hostOsFamily(): string {
  switch (process.platform) {
    case "win32": return "windows";
    case "darwin": return "macos";
    case "linux": return "linux";
    default: return process.platform;
  }
}

// ---------------------------------------------------------------------------
// Host facts
// ---------------------------------------------------------------------------
//
// The GPU and display can only be read by actually rendering something, so they come from a
// one-off no-persona launch. That is expensive (a browser start), so the result is cached on
// disk and keyed by the engine binary — a different build may report a different GL string.

interface CachedHost extends HostFacts { _exe: string; _at: string; }

function hostCachePath(): string {
  return join(cacheDir(), "host.json");
}

export function readCachedHost(exe: string): HostFacts | null {
  try {
    const c = JSON.parse(readFileSync(hostCachePath(), "utf8")) as CachedHost;
    if (c._exe !== exe) return null;               // different binary: re-measure
    // Host hardware does not change often, but a monitor or driver swap does; a month is a
    // reasonable compromise between staleness and paying for a launch every run.
    if (Date.now() - Date.parse(c._at) > 30 * 86_400_000) return null;
    return c;
  } catch {
    return null;
  }
}

export function writeCachedHost(exe: string, facts: HostFacts): void {
  try {
    writeFileSync(hostCachePath(), JSON.stringify({ ...facts, _exe: exe, _at: new Date().toISOString() }, null, 1));
  } catch {
    /* a cache write failure must never fail a launch */
  }
}

/** The in-page probe, as an expression string. Reads the REAL GPU and display — this launch
 *  applies no persona, so what it sees is the host itself. */
const PROBE_EXPR = `(() => {
  let renderer = "";
  try {
    const gl = document.createElement("canvas").getContext("webgl");
    const dbg = gl && gl.getExtension("WEBGL_debug_renderer_info");
    if (gl && dbg) renderer = String(gl.getParameter(dbg.UNMASKED_RENDERER_WEBGL));
  } catch (e) { /* headless with no GPU: leave blank, scored as unknown rather than guessed */ }
  return {
    renderer: renderer,
    screen_width: screen.width,
    screen_height: screen.height,
    device_pixel_ratio: devicePixelRatio,
    hardware_concurrency: navigator.hardwareConcurrency,
    device_memory: navigator.deviceMemory,
  };
})()`;

interface ProbeResult {
  renderer: string;
  screen_width: number;
  screen_height: number;
  device_pixel_ratio: number;
  hardware_concurrency: number;
  device_memory?: number;
}

/** Read the host's real GPU/display by rendering in the engine itself, with no persona applied. */
export async function measureHost(
  launchFn: (opts: Record<string, unknown>) => Promise<{ newContext: () => Promise<{ newPage: () => Promise<unknown> }>; close: () => Promise<void> }>,
  exe: string,
  browserMajor: number,
): Promise<HostFacts> {
  const cached = readCachedHost(exe);
  if (cached) return cached;

  const browser = await launchFn({ executablePath: exe, headless: true, quiet: true });
  try {
    const ctx = await browser.newContext();
    const page = (await ctx.newPage()) as {
      goto: (u: string, o?: unknown) => Promise<unknown>;
      evaluate: (expr: string) => Promise<ProbeResult>;
      close: () => Promise<void>;
    };
    // about:blank is enough: WebGL and screen do not need an origin.
    await page.goto("about:blank");
    // Passed as an EXPRESSION STRING, not a closure: this body runs in the browser, and typing
    // it here would mean adding the DOM lib to a Node-only SDK's tsconfig just to describe code
    // that never executes in Node.
    const facts = await page.evaluate(PROBE_EXPR);
    const out: HostFacts = {
      os_family: hostOsFamily(),
      browser_major: browserMajor,
      gpu_vendor: gpuVendorClass(facts.renderer),
      screen_width: facts.screen_width,
      screen_height: facts.screen_height,
      device_pixel_ratio: facts.device_pixel_ratio,
      hardware_concurrency: facts.hardware_concurrency,
      device_memory: facts.device_memory,
    };
    writeCachedHost(exe, out);
    return out;
  } finally {
    await browser.close().catch(() => {});
  }
}

// ---------------------------------------------------------------------------
// Service client
// ---------------------------------------------------------------------------

function authHeaders(licenseKey: string | undefined): Record<string, string> {
  const key = licenseKey || process.env.CLEARCOTE_LICENSE_KEY || "";
  if (!key) {
    throw new Error(
      "the profile library needs a license key — pass licenseKey or set CLEARCOTE_LICENSE_KEY " +
        "(profiles are served to PRO licenses with an active subscription)",
    );
  }
  return { authorization: `Bearer ${key}` };
}

/** Candidate index for a described host. Cached briefly so repeated launches do not re-ask. */
export async function fetchIndex(host: HostFacts, opts: ProfileSourceOptions = {}): Promise<ProfileIndexEntry[]> {
  const os = (opts.os ?? host.os_family).toLowerCase();
  const cachePath = join(cacheDir(), `index-${os}-${host.browser_major}.json`);

  if (opts.offline) {
    if (!existsSync(cachePath)) {
      throw new Error(`offline: no cached profile index for ${os}/${host.browser_major}`);
    }
    return JSON.parse(readFileSync(cachePath, "utf8")) as ProfileIndexEntry[];
  }

  const url = `${apiBase(opts.apiBase)}/api/v1/profiles?os=${encodeURIComponent(os)}&major=${host.browser_major}`;
  const res = await fetch(url, { headers: authHeaders(opts.licenseKey) });
  if (!res.ok) {
    // Serve stale rather than fail the launch — a cached index is far better than no persona.
    if (existsSync(cachePath)) return JSON.parse(readFileSync(cachePath, "utf8")) as ProfileIndexEntry[];
    const body = await res.text().catch(() => "");
    throw new Error(`profile index request failed (${res.status}): ${body.slice(0, 200)}`);
  }
  const json = (await res.json()) as { candidates: ProfileIndexEntry[] };
  try { writeFileSync(cachePath, JSON.stringify(json.candidates)); } catch { /* non-fatal */ }
  return json.candidates ?? [];
}

/** Fetch one profile by id, preferring the on-disk copy — a repeat costs no distinct budget,
 *  but not spending a request at all is better still. */
export async function fetchProfile(
  id: string,
  opts: ProfileSourceOptions = {},
): Promise<Record<string, unknown>> {
  const cachePath = join(cacheDir(), `profile-${id}.json`);
  if (existsSync(cachePath)) {
    try { return JSON.parse(readFileSync(cachePath, "utf8")) as Record<string, unknown>; } catch { /* refetch */ }
  }
  if (opts.offline) throw new Error(`offline: profile ${id} is not cached`);

  const res = await fetch(`${apiBase(opts.apiBase)}/api/v1/profiles?id=${encodeURIComponent(id)}`, {
    headers: authHeaders(opts.licenseKey),
  });
  if (!res.ok) {
    const body = await res.text().catch(() => "");
    throw new Error(`profile fetch failed (${res.status}): ${body.slice(0, 200)}`);
  }
  const json = (await res.json()) as { profile: Record<string, unknown> };
  try { writeFileSync(cachePath, JSON.stringify(json.profile)); } catch { /* non-fatal */ }
  return json.profile;
}

/** Resolve a persona end to end: index -> score against this host -> fetch the winner. */
export async function resolveAutoProfile(
  host: HostFacts,
  opts: ProfileSourceOptions = {},
): Promise<{ profile: Record<string, unknown>; selection: Selection }> {
  const index = await fetchIndex(host, opts);
  const selection = selectProfile(index, host, opts);
  const profile = await fetchProfile(selection.entry.id, opts);
  return { profile, selection };
}

File: sdk/python/clearcote/_profilesource.py
"""Fetching personas from the licensed profile service, and caching them locally.

Port of the Node SDK's ``profilesource.ts``.

Selection runs CLIENT-SIDE (see :mod:`_profilelib`) because it needs facts only this machine has:
the real GPU, the real display, the engine's actual Chromium major. The server cannot see any of
those, so it returns candidates for a described host and the SDK picks among them.

The service caps DISTINCT profiles per license per day, and re-fetching one already held is free.
That is why this module caches aggressively on disk: a sticky identity must resolve to the same
profile every launch without spending budget or needing the network.
"""

from __future__ import annotations

import json
import os
import sys
import urllib.error
import urllib.request
from typing import Any, Optional

from ._profilelib import gpu_vendor_class

DEFAULT_API_BASE = "https://www.clearcotelabs.com"

#: In-page probe reading the REAL GPU and display — this launch applies no persona.
PROBE_EXPR = """(() => {
  let renderer = "";
  try {
    const gl = document.createElement("canvas").getContext("webgl");
    const dbg = gl && gl.getExtension("WEBGL_debug_renderer_info");
    if (gl && dbg) renderer = String(gl.getParameter(dbg.UNMASKED_RENDERER_WEBGL));
  } catch (e) {}
  return {
    renderer: renderer,
    screen_width: screen.width,
    screen_height: screen.height,
    device_pixel_ratio: devicePixelRatio,
    hardware_concurrency: navigator.hardwareConcurrency,
    device_memory: navigator.deviceMemory,
  };
})()"""


def _api_base(base: Optional[str] = None) -> str:
    return (
        base
        or os.environ.get("CLEARCOTE_PROFILE_API")
        or os.environ.get("CLEARCOTE_LICENSE_API")
        or DEFAULT_API_BASE
    ).rstrip("/")


def profile_cache_dir() -> str:
    """``<CLEARCOTE_PROFILE_DIR>/library``, created on demand."""
    root = os.environ.get("CLEARCOTE_PROFILE_DIR") or os.path.join(
        os.path.expanduser("~"), ".clearcote", "profiles"
    )
    d = os.path.join(root, "library")
    os.makedirs(d, exist_ok=True)
    return d


def host_os_family() -> str:
    """Map the Python platform onto the ``os_family`` values the corpus stores."""
    if sys.platform.startswith("win"):
        return "windows"
    if sys.platform == "darwin":
        return "macos"
    if sys.platform.startswith("linux"):
        return "linux"
    return sys.platform


def _auth_headers(license_key: Optional[str]) -> dict[str, str]:
    key = license_key or os.environ.get("CLEARCOTE_LICENSE_KEY") or ""
    if not key:
        raise ValueError(
            "the profile library needs a license key — pass license_key or set "
            "CLEARCOTE_LICENSE_KEY (profiles are served to PRO licenses with an active "
            "subscription)"
        )
    return {"Authorization": f"Bearer {key}"}


def _get_json(url: str, headers: dict[str, str], timeout: int = 30) -> Any:
    req = urllib.request.Request(url, headers=headers)
    with urllib.request.urlopen(req, timeout=timeout) as resp:  # noqa: S310 - fixed https base
        return json.loads(resp.read().decode("utf-8"))


def fetch_index(
    host: dict[str, Any],
    license_key: Optional[str] = None,
    api_base: Optional[str] = None,
    offline: bool = False,
    os_override: Optional[str] = None,
) -> list[dict[str, Any]]:
    """Candidate index for a described host. Cached so repeated launches do not re-ask."""
    os_family = (os_override or host.get("os_family") or "").lower()
    major = host.get("browser_major")
    cache_path = os.path.join(profile_cache_dir(), f"index-{os_family}-{major}.json")

    if offline:
        if not os.path.isfile(cache_path):
            raise ValueError(f"offline: no cached profile index for {os_family}/{major}")
        with open(cache_path, "r", encoding="utf-8") as f:
            return json.load(f)

    url = f"{_api_base(api_base)}/api/v1/profiles?os={os_family}&major={major}"
    try:
        data = _get_json(url, _auth_headers(license_key))
    except Exception:
        # Serve stale rather than fail the launch — a cached index beats no persona.
        if os.path.isfile(cache_path):
            with open(cache_path, "r", encoding="utf-8") as f:
                return json.load(f)
        raise

    candidates = data.get("candidates") or []
    try:
        with open(cache_path, "w", encoding="utf-8") as f:
            json.dump(candidates, f)
    except OSError:
        pass
    return candidates


def fetch_profile(
    pid: str,
    license_key: Optional[str] = None,
    api_base: Optional[str] = None,
    offline: bool = False,
) -> dict[str, Any]:
    """Fetch one profile by id, preferring the on-disk copy.

    A repeat costs no distinct budget at the service, but not spending a request at all is
    better still.
    """
    cache_path = os.path.join(profile_cache_dir(), f"profile-{pid}.json")
    if os.path.isfile(cache_path):
        try:
            with open(cache_path, "r", encoding="utf-8") as f:
                return json.load(f)
        except (OSError, ValueError):
            pass  # refetch

    if offline:
        raise ValueError(f"offline: profile {pid} is not cached")

    data = _get_json(f"{_api_base(api_base)}/api/v1/profiles?id={pid}", _auth_headers(license_key))
    profile = data.get("profile") or {}
    try:
        with open(cache_path, "w", encoding="utf-8") as f:
            json.dump(profile, f)
    except OSError:
        pass
    return profile


# --- host facts -------------------------------------------------------------
# GPU and display can only be read by rendering, so they come from a one-off no-persona launch.
# That is expensive, so the result is cached on disk, keyed by the engine binary — a different
# build may report a different GL string.

_HOST_CACHE_MAX_AGE = 30 * 86400


def _host_cache_path() -> str:
    return os.path.join(profile_cache_dir(), "host.json")


def read_cached_host(exe: str) -> Optional[dict[str, Any]]:
    import time

    try:
        with open(_host_cache_path(), "r", encoding="utf-8") as f:
            c = json.load(f)
        if c.get("_exe") != exe:
            return None  # different binary: re-measure
        if time.time() - c.get("_at", 0) > _HOST_CACHE_MAX_AGE:
            return None
        return c
    except (OSError, ValueError):
        return None


def write_cached_host(exe: str, facts: dict[str, Any]) -> None:
    import time

    try:
        with open(_host_cache_path(), "w", encoding="utf-8") as f:
            json.dump({**facts, "_exe": exe, "_at": time.time()}, f)
    except OSError:
        pass  # a cache write failure must never fail a launch


def measure_host(launch_fn, exe: str, browser_major: int) -> dict[str, Any]:
    """Read the host's real GPU/display by rendering in the engine itself, no persona applied."""
    cached = read_cached_host(exe)
    if cached:
        return cached

    browser = launch_fn(executable_path=exe, headless=True, quiet=True)
    try:
        page = browser.new_context().new_page()
        page.goto("about:blank")  # WebGL and screen need no origin
        facts = page.evaluate(PROBE_EXPR)
        out = {
            "os_family": host_os_family(),
            "browser_major": browser_major,
            "gpu_vendor": gpu_vendor_class(facts.get("renderer")),
            "screen_width": facts.get("screen_width"),
            "screen_height": facts.get("screen_height"),
            "device_pixel_ratio": facts.get("device_pixel_ratio"),
            "hardware_concurrency": facts.get("hardware_concurrency"),
            "device_memory": facts.get("device_memory"),
        }
        write_cached_host(exe, out)
        return out
    finally:
        try:
            browser.close()
        except Exception:  # noqa: BLE001
            pass
